### ISSUE TITLE
`no-unused-vars` is incorrectly raised when function name matches parameter name

### ISSUE THREAD

**Issue by sokra** [eslint/eslint#2363]
The code snippet:

``` js
function a(a) { return a; }
a();
```

incorrectly complain about the parameter `a` being unused. It seem to prefer the reference to the funcion name instead of the parameter.

Propably caused by the order [in `block-scoped-var.js`](https://github.com/eslint/eslint/blob/d386abb47c76f457382b37cde969c2d7942d7464/lib/rules/block-scoped-var.js#L213-L214).

**Comment by nzakas** [comment]
Thanks, we'll take a look.

**Comment by jrvidal** [comment]
I think the `a` reference in the `return a` statement is being incorrectly classified as the function name by [`isSelfReference`](https://github.com/eslint/eslint/blob/ebbbff45adb56cf6c4a177bbe9abc306e6bc4258/lib/rules/no-unused-vars.js#L69-L76). It looks like it doesn't account for shadowing.

Would it make sense to extract the shadowing checks as a common tool?

**Comment by nzakas** [comment]
Perhaps. We do have a lot of repeating logic across rules, not just related to scoping. I'm definitely open to suggestions to eliminate duplicate code.

**Comment by jrvidal** [comment]
I've been digging a little bit about this one. Some previous discussions can be found in #2095, df9b2549438b98e68ff12ef049c5ccf54c12a58d.

Here are a few test cases related to `no-unused-vars` and recursive functions:

1. The original report:
    ```javascript
    "use strict";

    function a(a) {
    // Incorrectly reports the parameter 'a' as unused
        return a;
    }
    
    a(1); // 1
    ```

1. An analogous case with an inner function:
    ```javascript
    "use strict";
    
    function a() {
        // Incorrectly reports the 'a' inner function as unused
        function a() {
            return 1;
        }
    
        return a();
    }
    
    a(); // 1
    ```

1. If a function expression is used, it only gets reported if it's named:
    ```javascript
    "use strict";
    
    var a = function () {
        // OK, not reported
        var a = 1;
    
        return a;
    };
    
    var b = function b() {
        // Incorrectly reports 'b' as unused
        var b = 1;
    
        return b;
    };

    var c = function d() {
        // Incorrectly reports 'd' as unused
        var d = 1;
    
        return d;
    };

    a(); // 1
    b(); // 1
    c(); // 1
    ```

1. On the other hand, function expressions are not checked for self-reference:
    ```javascript
    "use strict";

    // Incorrectly? considers 'a' as used
    var a = function () {
        return a();
    };
    ```

1. As a _very_ contrived example, recursive function declarations are not always detected eiter:
    ```javascript
    "use strict";

    // Incorrectly? considers 'foo' as used
    function foo() {
        return function inner() {
            if (Math.random() < 0.5) {
                return foo();
            } else {
                return 1;
            }
        };
    }
    ```

===

For now, I'm trying to find a fix for the "false positives" in 1, 2 and 3. I am not sure whether detecting all "missed negatives" (undetected unused vars) statically is even feasible/desirable.

**Comment by jrvidal** [comment]
Actually there are [tests](https://github.com/eslint/eslint/blob/2f97e89f1004668b307ffbfa415969d6da0797ca/tests/lib/rules/no-unused-vars.js#L54-55) that consider case 5 above as valid. Should we consider changing this behavior?

**Comment by nzakas** [comment]
The recursive stuff is hard to track down, especially in the case of anonymous functions. I'd say start with the first three cases and focus on getting those fixed.
